tools/src/wyvern/tools/parsing/coreparser/arch/DeclCheckVisitor.java

```diff
@@ -0,0 +1,224 @@
+            }
+
+        }
+        return super.visit(node, data);
+    }
+
+    public Object visit(ASTPortDecl node, Object data) {
+        portdecls.put(node.getPort(), node);
+        return super.visit(node, data);
+    }
+
+    public Object visit(ASTEntryPointDecl node, Object data) {
+        entrypoints.put(node.getName(), node.getAction());
+        return super.visit(node, data);
+    }
+
+    public Object visit(ASTComponentDecl node, Object data) {
+        String name = node.getName();
+        String type = node.getType();
+        if (!componentTypes.containsKey(type)) {
+            // component type not declared
+            ToolError.reportError(ErrorMessage.ARCH_TYPE_NOT_DEFINED,
+                    node.getLocation(), "component", type);
+        }
+        if (components.putIfAbsent(name, node) != null) {
+            // duplicate members
+            ToolError.reportError(ErrorMessage.DUPLICATE_MEMBER_NAMES,
+                    node.getLocation(), name);
+        }
+        return super.visit(node, data);
+    }
+
+    public Object visit(ASTConnectorDecl node, Object data) {
+        String name = node.getName();
+        String type = node.getType();
+        if (!connectorTypes.contains(type)) {
+            // connector type not declared
+            ToolError.reportError(ErrorMessage.ARCH_TYPE_NOT_DEFINED,
+                    node.getLocation(), "connector", type);
+        }
+        if (connectors.putIfAbsent(name, type) != null) {
+            // duplicate members
+            ToolError.reportError(ErrorMessage.DUPLICATE_MEMBER_NAMES,
+                    node.getLocation(), name);
+        }
+        return super.visit(node, data);
+    }
+
+    public Object visit(ASTComponentTypeDecl node, Object data) {
+        if (componentTypes.containsKey(node.getTypeName())) {
+            // type of this name has already been declared
+            ToolError.reportError(ErrorMessage.DUPLICATE_TYPE_DEFINITIONS,
+                    node.getLocation(), node.getTypeName());
+        }
+        node.collectPorts();
+        if (!node.checkModule(state)) {
+            // module def not found HANDLED IN CHECKMODULE
+            return super.visit(node, data);
+        } else {
+            componentTypes.put(node.getTypeName(), node);
+        }
+        return super.visit(node, data);
+    }
+
+    public Object visit(ASTConnectorTypeDecl node, Object data) {
+        if (connectorTypes.contains(node.getTypeName())) {
+            // type of this name has already been declared
+            ToolError.reportError(ErrorMessage.DUPLICATE_TYPE_DEFINITIONS,
+                    node.getLocation(), node.getTypeName());
+        }
+        node.collectVals();
+        if (!node.checkModule(state)) {
+            // module not found HANDLED IN CHECKMODULE
+            return super.visit(node, data);
+        } else {
+            connectorTypes.add(node.getTypeName());
+        }
```

Comment:
I would change this if-else block to the following:

if (node.checkModule(state)) {
&nbsp;&nbsp;&nbsp;&nbsp;connectorTypes.add(node.getTypeName());
}